catalog/app/containers/Bucket/requests/packageUpload.ts

```diff
@@ -0,0 +1,258 @@
+    bucket: string
+    key: string
+    version: string
+  }
+  size: number
+}
+
+interface ManifestBodyBase {
+  message: string
+  meta: JsonValue
+  registry: string
+  workflow?: string | null
+}
+
+interface ManifestBodyCreate extends ManifestBodyBase {
+  contents: FileUpload[]
+  name: string
+}
+
+interface ManifestBodyCopy extends ManifestBodyBase {
+  name: string
+  parent: {
+    top_hash: string
+    registry: string
+    name: string
+  }
+}
+
+interface ManifestBodyWrap extends ManifestBodyBase {
+  dst: {
+    registry: string
+    name: string
+  }
+  entries: FileEntry[]
+}
+
+type ManifestBody = ManifestBodyCreate | ManifestBodyWrap | ManifestBodyCopy
+
+const ENDPOINT_CREATE = '/packages'
+
+const ENDPOINT_UPDATE = '/packages'
+
+const ENDPOINT_COPY = '/packages/promote'
+
+const ENDPOINT_WRAP = '/packages/from-folder'
+
+type Endpoint =
+  | typeof ENDPOINT_CREATE
+  | typeof ENDPOINT_UPDATE
+  | typeof ENDPOINT_COPY
+  | typeof ENDPOINT_WRAP
+
+// TODO: reuse it from some other place, don't remember where I saw it
+interface ManifestHandleTarget {
+  bucket: string
+  name: string
+}
+
+interface ManifestHandleSource extends ManifestHandleTarget {
+  revision: string
+}
+
+interface BasePackageParams {
+  message: string
+  meta: JsonValue
+  workflow: workflows.Workflow
+}
+
+interface CreatePackageParams extends BasePackageParams {
+  contents: FileUpload[]
+  target: {
+    bucket: string
+    name: string
+  }
+}
+
+interface UpdatePackageParams extends BasePackageParams {
+  contents: FileUpload[]
+  source: {
+    bucket: string
+    name: string
+  }
+}
+
+interface CopyPackageParams extends BasePackageParams {
+  source: ManifestHandleSource
+  target: ManifestHandleTarget
+}
+
+interface WrapPackageParams extends BasePackageParams {
+  entries: FileEntry[]
+  source: string
+  target: ManifestHandleTarget
+}
+
+interface Response {
+  top_hash: string
+}
+
+// FIXME: this is copypasted from PackageDialog -- next time we need to TSify utils/APIConnector properly
+interface ApiRequest {
+  <O>(opts: {
+    endpoint: string
+    method?: 'GET' | 'PUT' | 'POST' | 'DELETE' | 'HEAD'
+    body?: {}
+  }): Promise<O>
+}
+
+const uploadManifest = (
+  req: ApiRequest,
+  endpoint: Endpoint,
+  body: ManifestBody,
+): Promise<Response> =>
+  req({
+    endpoint,
+    method: 'POST',
+    body,
+  })
+
+const getMetaValue = (value: unknown, optSchema: JsonSchema) =>
+  value
+    ? pipeThru(value || {})(
+        makeSchemaDefaultsSetter(optSchema),
+        R.toPairs,
+        R.filter(([k]) => !!k.trim()),
+        R.fromPairs,
+        R.when(R.isEmpty, () => undefined),
+      )
+    : undefined
+
+const getWorkflowApiParam = R.cond([
+  [R.equals(workflows.notAvailable), R.always(undefined)],
+  [R.equals(workflows.notSelected), R.always(null)],
+  [R.T, R.identity],
+]) as (
+  slug: typeof workflows.notAvailable | typeof workflows.notSelected | string,
+) => string | null | undefined
+
+const createPackage = (
```

Comment:
i think we can reduce duplication and complexity by placing the request invocation inline, like this:
```typescript
export function useCreatePackage() {
  const req: ApiRequest = APIConnector.use()
  return React.useCallback(
    (
      { contents, message, meta, target, workflow }: CreatePackageParams,
      schema: JsonSchema,
    ) =>
      req<Response, CreateRequestBody>({
        endpoint: ENDPOINT_CREATE,
        method: 'POST',
        body: {
          name: target.name,
          registry: `s3://${target.bucket}`,
          message,
          contents,
          meta: getMetaValue(meta, schema),
          workflow: getWorkflowApiParam(workflow.slug),
        }), // or even just cast this object to CreateResponseBody
    [req],
  )
}
```